# 0. Abstract and Outline


The purpose of this document is to give a gentle introduction to supervised learning and linear regression in a nutshell. The application of linear regression to predicting the hotel room prices. Linear regression is arguably the simplest model used in regression and supervised learning. Besides its simplicity and scalability, another important advantage of linear regression is that it is well interpretable. Both the modeling details, the case study, and the  Python code are included in this document.

- In [Section 1](#section_1), we introduce the basic framework for supervised learning.
- In [Section 2](#section_2), we introduce the linear regression model and its estimation method.
- In [Section 3](#section_3), we present the case study of hotel room pricing.
- In [Section 4](#section_4), we introduce how to use linear regression for hotel room price prediction.
- In [Section_5](#section_5), we discuss how to evaluate and understand linear regression models in general.

In this and other notebooks of this course, we will use $\mathbb R^p$ to denote the $p-$dimensional real Euclidean space ([Wiki Page link](https://en.wikipedia.org/wiki/Euclidean_space)), $\mathbb P[A]$ to denote the probability of an event $A$, and $\mathbb E[X]$ to denote the expectation of a random variable $X$.

### Heads Up 

Starting from this session, we will inevitablly use a lot of **mathematical notations**, this is because:

- Mathematics enables us to represent some concepts, ideas, and implementations **neatly**, **succintly**, and **elegantly**, which would otherwise be impossible or very cumbersome.
- The mathematical notations introduced in this course are the standard **language** in the data science/data analytics community. Familiarizing ourselves with them is important for us to communicate with other professionals in this community.

Importantly, I would emphasize that, although the mathematical notations look intimidating, you will find that they are so **natural** and **helpful** once you understand the logic behind them.

<a id='section_1'></a>
# 1. Supervised Learning in a Nutshell

Supervised learning refers to a wide variety of machine learning task through which one seeks to identify the underlying relationship between the ($p-$dimensional) feature $X=(X_1,X_2,...,X_p)'\in\mathbb R^p$ ( ' is the transpose operation for a matrix ) and label $Y\in\mathbb R$. 

In supervised learning, $Y$ is the key variable of interest, whose value is so important for decision making that we wish to predict as accurate as possible. Typical examples of $Y$ in applications include:

- The daily active users (DAU) and daily usage time per user of an app like Tiktok.
- The amount of money a user uses to purchase products on Amazon per year.
- The wage of a worker in a certain industry segment.
- Etc.

In the literature and practical applications, $Y$ also has some other names such as dependent variable, response variable, target variable, output variable, outcome variable, etc.

The feature vector $X$ contains the variable(s) that could be used to explain and/or predict $Y$. The key value of the features $X$ lies in their capability to accurately predict $Y$. Typical examples of $X$ in applications include:

- The user demographic data of an App.
- The user behavior data of an App.
- The characteristics of a product.
- Etc.

In different contexts, $X$ may have some other names such as independent variable, covariate, predictor, explanatory variable, input variable, data, etc.

To be more specific, we seek to build a learner (aka model) $\hat f(\cdot)$ based on the training data set of sample size $n$,
$$\mathcal D:=\{(Y_i,X_i):Y_i\in\mathbb R,X_{i}\in\mathbb R^p,1\le i\le n\}$$ 
that predicts the ourcome $Y$ from feature $X$ of the following form:

$$Y_i\approx\hat Y_i:=\hat f(X_i)\mbox{ for all }i=1,2,...,n
$$

Here, $\approx$ refers to "as close as possible", or some kind of error is minimized. In this course, we use the notation $\hat{}$ to represent what is generated from data. At a high level, we train our model $\hat f(\cdot)$ to make as few mistakes as possible (on the training set):

$$\frac{1}{n}\sum_{i=1}^n \mathcal L(\hat f(X_i),Y_i)\mbox{ is minimized, where }\mathcal L(\cdot,\cdot)\mbox{ is some error measure between the predicted value}\hat f(X_i)\mbox{ and true value }Y_i.
$$
We will discuss some specific loss functions later in this course.

With supervised learning, we generally aim at achieving the following by identifying a good candidate of $\hat f(\cdot)$:

- Understand which of the features in $X$ actually have some (significant) relationships with $Y$.
- Characterize whether the association/correlation between $X$ and $Y$ is positive or negative.
- Identify the weight of each feature in the vector $X$ on the outcome $Y$.

We now introduce some examples of supervised learning:

- Use movie features (genre, actors, average rating, length, etc.) and user features (age, gender, location, ratings given to other films, etc.) to predict the rating a user will give to a movie s/he has never watched before.
- Use user past behaviors to predict his/her total app time of Kwai tomorrow.
- Use email texts and figures to predict whether it is a spam.
- Use CT scan images to predict whether the patient has lung cancer.

There are two types of supervised learning, regression and classification. For regression, $Y$ is a real-valued continuous variable such as rating, app time, price, etc. The commonly used loss/error functions in regression are squared error $\mathcal L(\hat f(X_i),Y_i)=(\hat f(X_i)-Y_i)^2$ and absolute error $\mathcal L(\hat f(X_i),Y_i)=|\hat f(X_i)-Y_i|$, etc. For classification, $Y$ a factor/categorical variable (i.e., taking discrete values, such as in binary classification, $Y\in\{0,1\}$). Examples $Y$ in classification problems include whether an email is a spam, whether a patient has cancer, and whether a customer will purchase a product, etc. Commonly used loss functions in classification problems include the zero-one loss $\mathcal L(\hat f(X_i),Y_i)=1\{Y_i\ne \hat f(X_i)\}=\begin{cases}
1,\mbox{ if }Y_i\ne \hat f(X_i)\\
0,\mbox{ otherwise},
\end{cases}
$.

It would be useful to preview and discuss the similarities and differences between supervised learning and unsupervised learning. From a similarity point of view, both approaches attempt to model data, address certain practical problems, and eventually deliver business and societal values. The key difference between supervised learning and unsupervised learning can be summarized as follows:

- Supervised learning aims to directly model the relationship between features and label, so that labels in new data can be accurately predicted given features.

- Unsupervised learning finds patterns/relationships/structures in data without being optimized to solve a particular predictive task (i.e., there is not label $Y$ for unsupervised learning).

<a id='section_2'></a>
# 2. Linear Regression Model and Estimation

Linear regression is arguably the simplest supervised learning model. The key assumption for linear regression is that the label $Y$ and feature $X$ has a linear relationship:

$$Y_i=\beta_0+\sum_{j=1}^p\beta_jX_{ij}+\epsilon_{i},\mbox{ for }i=1,2,...,n$$

where $\beta_j$ is the coefficient for feature $j$ and $\epsilon_i$ is a random noise which is independently and identically distributed (called i.i.d. in statistics text books, see this [Wiki Page](https://en.wikipedia.org/wiki/Independent_and_identically_distributed_random_variables)) with mean 0 and finite variance ($\mathbb E[\epsilon_i]=0$ and $\mathbb E[\epsilon_i^2]<+\infty$). The question is now reduced to estimate the value of $\beta=(\beta_0,\beta_1,...,\beta_p)$. 

We denote $\hat\beta=(\hat\beta_0,\hat\beta_1,...\hat\beta_p)$ as the estimation of $\beta$ fitted on the data set $$\mathcal D:=\{(Y_i,X_i):Y_i\in\mathbb R,X_{i}\in\mathbb R^p,1\le i\le n\}$$ 
The most commonly adopted method to fit a linear regression model is ordinary least squares (OLS, see this [Wiki Page](https://en.wikipedia.org/wiki/Ordinary_least_squares)):

$$\hat\beta=\mbox{argmin}_{\beta}\sum_{i=1}^n(Y_i-\beta_0-\sum_{j=1}^p\beta_jX_{ij})^2$$

As is clear from the definition, the OLS estimate $\hat\beta$ minimizes the sum of squared errors (SSE), which is very easy to compute. In other words, the loss function is the squared loss: $\mathcal L(Y,\hat Y)=(Y-\hat Y)^2$. An important property of OLS is that it penalizes more on larger errors due to the squared error. In general, the squred error can also be used to predict the conditional mean of the label $Y$ given feature $X$, $f(X):=\mathbb E [Y|X]$. The estimator of $\hat f(\cdot)$ is given by:

$$\hat f(\cdot)=\mbox{argmin}_{f(\cdot)}\sum_{i=1}^n(Y_i-f(X_i))^2$$

To conclude this section, we remark that, the estimated coefficient for feature $j$, $\hat \beta_j$ ($1\le j\le p$) can be interpreted as "one unit of change in $X_j$ is associated with $\hat\beta_j$ units of change in $Y$". We will get back to the interpretation problem of linear regression models later in this notebook.

<a id='section_3'></a>
# 3. Case Study: Predicting Hotel Room Prices

Needless to say, pricing is a key decision for hotel managers. In this case, we study how to use machine learning (more specifically, linear regression) to predict the hotel room prices. There are 3 potential approaches to predict hotel room prices:

- Solution 1: Design a pricing rule based on prior knowledge. For example, we simply predict that, if the hotel is in city 1, then the price is 500; if the hotel is in city 2, then the price is 1000; if the hotel is in city 3, then the price is 1500. 
- Solution 2: Set the price as the average room price of the hotels close to the focal one (e.g., within 1km in distance).
- Solution 3: Identify which features of the hotel room are associated with higher prices and set the prices accordingly.

The above 3 solutions have their respective pros and cons, which we list as follows:

<table style="width:100%" align="center">
  <tr>
    <th></th>
    <th>Advanrages</th> 
    <th>Disadvantages</th>
  </tr>
  <tr>
    <td>Solution 1</td>
    <td>Data not required</td>
    <td>Vulnerable to wrong assumptions and not adaptive to new data/information</td>
  </tr>
  <tr>
    <td>Solution 2</td>
    <td>Labeled data not required</td>
    <td>Vulnerable to false assumptions and hard to adapt to new data/information</td>
  </tr>
    <tr>
    <td>Solution 3</td>
    <td>Directly minimizing an error metric and easily adaptive to new data/information</td>
    <td>Large labeled data set required</td>
  </tr>
</table>

In the following, we will study Solution 3 and propose a linear regression model to predict the hotel room prices. The main advantage of this approach is its scalability and adaptiveness to new data/information.

Let's first load the data into Python. To begin with, we import necessary packages.

In [1]:
import sys 
import numpy as np
import pandas as pd
import statsmodels as sm
import sklearn
import scipy as sp
%matplotlib inline 
import matplotlib.pyplot as plt
import math

In [2]:
# Load the hotel room price data into Python as a DataFrame.
df = pd.read_csv('room_pricing.csv',delimiter=',',index_col=0)
# Take a look at the first 10 rows, using the head() method of Pandas DataFrame.
df.head(10)

city  is_within_two_years  num_rating  rating_location  rating_service  \
index                                                                          
1       C2                    0         307         4.516447        4.614064   
2       C1                    0         322         4.375212        4.556914   
3       C2                    0         185         4.305307        4.536324   
4       C2                    1         470         4.335570        4.596367   
5       C2                    0         328         4.582442        4.414963   
6       C2                    0         402         4.373378        4.376965   
7       C3                    0         301         4.592301        4.557452   
8       C2                    0          91         4.410937        4.409150   
9       C2                    0         579         4.215112        4.471941   
10      C2                    0         103         4.328122        4.527167   

       rating_facility  POI_firm  POI_school  POI_travel      room   price  
index                                                                       
1             4.311731       167         144         229  business  1307.0  
2             4.638756       120          32         195  business   689.0  
3             4.652649       121          75         178  business  1276.0  
4             4.132112       209         164          14  standard   763.0  
5             4.694790        32          98         126  business  1090.0  
6             4.468706       220         179         200  business   835.0  
7             4.517460       251         251         221  standard   615.0  
8             4.231240       218         179         135  business   892.0  
9             4.632835         7         104          17    luxury  1517.0  
10            4.528234        31          75         235    luxury  1747.0

Each row represents one hotel room. As we can see, the features of this data are:

- **city**: Which city the hotel is in, taking values of "C1", "C2" or "C3";
- **is_within_two_years**: Whether the hotel was built within 2 years;
- **num_rating**: The number of ratings on a third-party OTA (online-travel-agency);
- **rating_location**: The average rating of this room on location;
- **rating_service**: The average rating of this room on service;
- **rating_facility**: The average rating of this room on facility;
- **POI_firm**: The number of POIs (point-of-interest) related to firms;
- **POI_school**: The number of POIs related to schools;
- **POI_travel**: The number of POIs related to tourist attractions;
- **room**: The room type, taking values of "business", "standard", or "luxury".

The label is the following varible:

- **price**: The price per night for the hotel room.

Let's first do some descriptive analysis for this data set.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6125 entries, 1 to 6125
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   city                 6125 non-null   object 
 1   is_within_two_years  6125 non-null   int64  
 2   num_rating           6125 non-null   int64  
 3   rating_location      6125 non-null   float64
 4   rating_service       6125 non-null   float64
 5   rating_facility      6125 non-null   float64
 6   POI_firm             6125 non-null   int64  
 7   POI_school           6125 non-null   int64  
 8   POI_travel           6125 non-null   int64  
 9   room                 6125 non-null   object 
 10  price                6125 non-null   float64
dtypes: float64(4), int64(5), object(2)
memory usage: 574.2+ KB


In [4]:
df.describe()

is_within_two_years   num_rating  rating_location  rating_service  \
count          6125.000000  6125.000000      6125.000000     6125.000000   
mean              0.186286   304.637224         4.361745        4.530634   
std               0.389369   172.970142         0.099924        0.100172   
min               0.000000     0.000000         4.032402        4.169993   
25%               0.000000   156.000000         4.293669        4.464106   
50%               0.000000   305.000000         4.361547        4.530918   
75%               0.000000   456.000000         4.427893        4.598160   
max               1.000000   599.000000         4.718097        4.894796   

       rating_facility     POI_firm   POI_school   POI_travel        price  
count      6125.000000  6125.000000  6125.000000  6125.000000  6125.000000  
mean          4.459063   128.598694   124.618776   129.079347  1065.137469  
std           0.141550    73.990788    73.473235    73.689154   410.600953  
min           3.910081     0.000000     0.000000     0.000000    51.000000  
25%           4.362661    64.000000    61.000000    65.000000   735.000000  
50%           4.457331   129.000000   123.000000   130.000000  1013.000000  
75%           4.553575   194.000000   187.000000   192.000000  1415.000000  
max           5.011693   254.000000   254.000000   254.000000  2099.000000

<a id='section_4'></a>
# 4. Linear Regression: Getting Started

To start with, we consider a simple linear regression model where **price** is the label and the average rating (the average of **rating_location**, **rating_service**, and **rating_facility**) is the feature.

$$\mbox{price}_i\approx \hat\beta_0+\hat\beta_1\mbox{average_rating}_i$$

In [5]:
# Create the feature a_rating, representing the average rating of the hotel room.
df['a_rating']=(df['rating_location']+df['rating_service']+df['rating_facility'])/3
df.head(10)

city  is_within_two_years  num_rating  rating_location  rating_service  \
index                                                                          
1       C2                    0         307         4.516447        4.614064   
2       C1                    0         322         4.375212        4.556914   
3       C2                    0         185         4.305307        4.536324   
4       C2                    1         470         4.335570        4.596367   
5       C2                    0         328         4.582442        4.414963   
6       C2                    0         402         4.373378        4.376965   
7       C3                    0         301         4.592301        4.557452   
8       C2                    0          91         4.410937        4.409150   
9       C2                    0         579         4.215112        4.471941   
10      C2                    0         103         4.328122        4.527167   

       rating_facility  POI_firm  POI_school  POI_travel      room   price  \
index                                                                        
1             4.311731       167         144         229  business  1307.0   
2             4.638756       120          32         195  business   689.0   
3             4.652649       121          75         178  business  1276.0   
4             4.132112       209         164          14  standard   763.0   
5             4.694790        32          98         126  business  1090.0   
6             4.468706       220         179         200  business   835.0   
7             4.517460       251         251         221  standard   615.0   
8             4.231240       218         179         135  business   892.0   
9             4.632835         7         104          17    luxury  1517.0   
10            4.528234        31          75         235    luxury  1747.0   

       a_rating  
index            
1      4.480748  
2      4.523627  
3      4.498093  
4      4.354683  
5      4.564065  
6      4.406350  
7      4.555737  
8      4.350443  
9      4.439963  
10     4.461174

We now run a linear regression using the sklearn package. See [Linear_Regression Documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html) for details. You may find a lot of useful built-in functions related to linear regression.

In [6]:
from sklearn.linear_model import LinearRegression as LinReg

# Create the feature and label for the regression model.
X = np.array(df['a_rating']).reshape([6125,1]) # X is the feature, 6125 rows by 1 column.
y = np.array(df['price']) # y is the label

# Fit a linear regression model with intercept (\beta_0 is not zero)
reg = LinReg(fit_intercept=True).fit(X, y)

# Print the coefficients:
print("The coefficient of average rating is: ",reg.coef_)
print("\n")
print("The intercept is: ",reg.intercept_)

The coefficient of average rating is:  [-30.74436557]


The intercept is:  1201.964681990667


So the fitted linear regression model is:

$$\mbox{price}_i\approx 1201.96-30.74\times\mbox{average_rating}_i$$

One could interpret the model results as:

- For a room whose average rating is 0, the average price is predicted as RMB 1202.0.
- One unit of increase in average rating is **Associated** with a decrease of RMB 30.74 in room price.

**Note:** Association/correlation is not equal to causality.

Another question we wish to address is how to evaluate whether this model is a good one. The next subsection is devoted to this question.

## 4.1. Regression Model Evaluation: $R^2$

A commonly used model evaluation metric is $R^2$. With the fitted model $\hat \beta=(\hat\beta_0,\hat\beta_1,...,\hat\beta_p)$, the fitted label of data point $i$ is given by

$$\hat Y_i=\hat\beta_0+\sum_{j=1}^p\hat\beta_j X_{ij}$$

We remark that for OLS, the fitted mean is the same as the sample mean:

$$\bar Y=\frac{1}{n}\sum_{i=1}^n Y_i=\frac{1}{n}\sum_{i=1}^n\hat Y_i.$$

The (in-sample) $R^2$ of the fitted model is given by

$$R^2=\frac{\sum_{i=1}^n(\hat Y_i-\bar Y)^2}{\sum_{i=1}^n(Y_i-\bar Y)^2}=\frac{\mbox{Sample Variance of }\hat Y_i}{\mbox{Sample Variance of }Y_i}=1-\frac{SSE}{\mbox{Sample Variance of }Y_i},$$
where $SSE$ is the sum of squared errors, 
$$SSE=\sum_{i=1}^n(Y_i-\hat Y_i)^2$$

As is clear from its definition, $R^2$ measures how well the model fits the data. More specifically, $R^2$ quantifies how much of the outcome sample variance can be "explained" by the fitted values. It is also the improvement (in label variance reduction) of the "best" linear regression model over the baseline model that ignores all the features. One should note that $R^2$ is universally interpretable across different models, but not a good metric to compare for different models. Furthermore, we use the same data to fit the model and to evaluate it (so, $R^2$ is an in-sample measure of fit). Hence, a model with a large (resp. small) $R^2$ may still be poor (resp. good).

The in-sample $R^2$ is upper-bounded by 1 and lower-bounded by 0 ($0\le R^2\le 1$). As a pictorial illustration, the following Figure (adapted from the book *[Elements of Statistical Learning](https://web.stanford.edu/~hastie/Papers/ESLII.pdf)*) illustrates the prediction of OLS linear regression $\hat Y$ is essentially a projection of $Y$ on the feature space. Hence, by the Pythagorean theorem, $|Y-\hat Y|^2+|\hat Y|^2=|Y|^2$, so $R^2=1-|Y-\hat Y|^2/|Y|^2=|\hat Y|^2/|Y|^2\in[0,1]$.

<img src="OLS.png" width=500>


Next, we introduce how to use the ```sklearn``` package to compute the $R^2$ of a linear regression model.

In [7]:
# print the R-squared of the linear regression model, using the score() function in the sklearn package.
print("The in-sample R-squared of the linear regression model is:", reg.score(X,y))

The in-sample R-squared of the linear regression model is: 2.438467587106974e-05


Let's then calculate the $R^2$ of a linear regression model with more features. Specifically, we will encode the average rating, city, and room type information into the feature vector. To this end, we first transform the categorical variables **City** and **room** into numeric ones. We introduce variable **c1** as the binary variable indicating whether the city is 'C1', **c2** as the binary variable indicating whether the city is 'C2', **rb** as the binary variable indicating whether the room type is 'business', and **rl** as the binary variable indicating whether the room type is 'luxury'. Hence, if **c1**=**c2**=0 the city is 'C3'. Likewise, if **rb**=**rl**=0, the room type is  standard.

In [8]:
df['c1'] = np.array(df['city']=='C1').astype(int)
# df['city']=='C1' returns an array of binary variables (True/False). astype(int) transforms True/False to 1/0.
df['c2'] = np.array(df['city']=='C2').astype(int)
df['rb'] = np.array(df['room']=='business').astype(int)
df['rl'] = np.array(df['room']=='luxury').astype(int)
df.head()

city  is_within_two_years  num_rating  rating_location  rating_service  \
index                                                                          
1       C2                    0         307         4.516447        4.614064   
2       C1                    0         322         4.375212        4.556914   
3       C2                    0         185         4.305307        4.536324   
4       C2                    1         470         4.335570        4.596367   
5       C2                    0         328         4.582442        4.414963   

       rating_facility  POI_firm  POI_school  POI_travel      room   price  \
index                                                                        
1             4.311731       167         144         229  business  1307.0   
2             4.638756       120          32         195  business   689.0   
3             4.652649       121          75         178  business  1276.0   
4             4.132112       209         164          14  standard   763.0   
5             4.694790        32          98         126  business  1090.0   

       a_rating  c1  c2  rb  rl  
index                            
1      4.480748   0   1   1   0  
2      4.523627   1   0   1   0  
3      4.498093   0   1   1   0  
4      4.354683   0   1   0   0  
5      4.564065   0   1   1   0

In [9]:
# Include the average rating, city, and room type information into the feature vector.
X1 = np.array(df[['a_rating','c1','c2','rb','rl']]).reshape([6125,5])

# Fit a linear regression model with intercept (\beta_0)
reg1 = LinReg(fit_intercept=True).fit(X1, y)

# Print the coefficients:
print("The coefficients are: ",reg1.coef_)
print("\n")
print("The intercept is: ",reg1.intercept_)
print("\n")
print("The (in-sampled) R-squared is: ",reg1.score(X1,y))

The coefficients are:  [ 159.91514749 -151.02792129  209.80869702  229.50844224  836.35322308]


The intercept is:  -95.01660658162245


The (in-sampled) R-squared is:  0.8888181843904622


You can play with the data on your own and you will find that as you add more variables into the model, $R^2$ will be **improved with diminishing returns**. In principle, you may not want to include all the features into the model, because more features require more data to fit the model. Furthermore, including too many varibles as the feature may also cause overfitting (see this [Wiki Page](https://en.wikipedia.org/wiki/Overfitting)) and bad performance on unseen data. In the next section, we will introduce how to give fair estimates for the performance of a linear regression model on unseen data and select the appropriate model accordingly.

<a id='section_5'></a>
# 5. More on Evaluating and Understanding Linear Regression

In this section, we will discuss the gold-standard of evaluating a linear regression model: the out-of-sample R-squared (OOS-$R^2$). Furthermore, we will dig deeper into understanding and interpreting the results from fitting a linear regression model.  

## 5.1. Out-of-Sample $R^2$

Recall that the in-sample $R^2$ measures the fitness between the linear regression model and the training data. However, we are more interested in understanding how well the model will perform on a **new** and unseen data set. To address this question, we introduce the notion of out-of-sample $R^2$, which evaluates the $R^2$ metric on a new data set not overlapping with the data used to fit the model. We refer to the data set for fitting the model as the training set, denoted as $\mathcal D_{tr}$ and the data set for evaluating the performance of the model as the testing set, denoted as $\mathcal D_{te}$. The sample size of the training set is denoted by $n_{tr}=|\mathcal D_{tr}|$ and the sample size of the testing set is denoted by $n_{te}=|\mathcal D_{te}|$. Computing the out-of-sample $R^2$ measure comprises of the following steps:

- **Step 1. Data Splitting.** Randomly split the entire data set $\mathcal D$ into non-overlapping training set $\mathcal D_{tr}$ and testing set $\mathcal D_{te}$. The sample size proportion is roughly $n_{tr}:n_{te}=7:3$ or $n_{tr}:n_{te}=6:4$.
- **Step 2. Model Training.** Train a linear regression model $\hat f(\cdot)$ (using OLS) on the training set $\mathcal D_{tr}$.
- **Step 3. Model Evaluation.** Compute the out-of-sample $R^2$ on the testing set $\mathcal D_{te}$:

$$\mbox{OOS-}R^2=1-\frac{\sum_{(X,Y)\in\mathcal D_{te}}(Y-\hat f(X))^2}{\sum_{(X,Y)\in\mathcal D_{te}}(Y-\bar Y_{tr})^2},\mbox{ where $\bar Y_{tr}:=\frac{1}{n_{tr}}\sum_{(X,Y)\in\mathcal D_{tr}}Y$ is the average outcome of the training set.}$$

Comparing the OOS-$R^2$ and the in-sample $R^2$ reveals a key property of the former metric that it is evaluated on a **different** data set. As such, we are not playing the roles of players and referees simultaneously, so the out-of-sample $R^2$ is a fair metric to evaluate the performance of a linear regression model on unseen data. One should note that higher in-sample $R^2$ does not necessarily imply higher out-of-sample $R^2$, due to overfitting. Furthermore, in theory, the out-of-sample $R^2$ is always upper-bounded by 1, but could be negative.

Next, we use OOS-$R^2$ to evaluate the performance of the two linear regression models built above.

In [10]:
# First, we import the train_test_split function to split the data into training and testing sets.
from sklearn.model_selection import train_test_split as tr_te_split

# Split the training and testing sets for the simple linear regression with one feature only.
X_train, X_test, y_train, y_test = tr_te_split(X, y, test_size=0.3) 
# The parameter test_size specifies the proportion of data in the testing set.

reg = LinReg(fit_intercept=True).fit(X_train, y_train)

# print the out-of-sample R-squared for the linear regression model.
print("The out-of-sample R-squared of the simple linear regression model is:", reg.score(X_test,y_test))

# Because of the randomness in training-testing sets split, the out-of-sample R-squared may be different for a 
# different split.

The out-of-sample R-squared of the simple linear regression model is: -0.005113519805040712


Next, we compute the out-of-sample $R^2$ for the more complex linear regression model with the average rating, city, and room type information in the feature vector.

In [11]:
# Split the training and testing sets for the more complex linear regression model with the average rating, city,
# and room type information in the feature vector.
X1_train, X1_test, y1_train, y1_test = tr_te_split(X1, y, test_size=0.3)

reg1 = LinReg(fit_intercept=True).fit(X1_train, y1_train)

# print the out-of-sample R-squared
print("The out-of-sample R-squared of the more complex linear regression model is:", reg1.score(X1_test,y1_test))

The out-of-sample R-squared of the more complex linear regression model is: 0.8875652382595896


As we can see, the second model not only has a high in-sample $R^2$, but also a high out-of-sample $R^2$.

## 5.2. Understanding Linear Regression Results

Estimated coefficients $\hat \beta$ are informative to characterize the association between the features and the label. More specifically, the sign of $\hat\beta_j$ characterizes whether the association between feature $X_j$ and the outcome $Y$ is positive or negative. The absolute value of the $\hat \beta_j$ measures how strong is the association. For the complex linear regression model, the estimated coefficients are:

In [33]:
# Store the coefficients
df_coeff = pd.DataFrame({
    'Features': ['a_rating','c1','c2','rb','rl'], 
    'Coefficients': reg1.coef_
})
#The estimated coefficients
df_coeff

Features  Coefficients
0  a_rating    165.032013
1        c1   -151.465499
2        c2    214.756788
3        rb    227.923234
4        rl    836.511011

Based on this estimation results, we have the following interpretation:

In [34]:
print("One unit of increase in average rating is associated with an RMB {:.2f} increase in the hotel room price.".format(df_coeff['Coefficients'][0]))
# {:.2f} means print 2 digits of the float number.
print("\n")
print("Compared with City 3, a hotel in City 1 is associated with an RMB {:.2f} decrease in the hotel room price.".format(-df_coeff['Coefficients'][1]))
print("\n")
print("Compared with City 3, a hotel in City 2 is associated with an RMB {:.2f} increase in the hotel room price.".format(df_coeff['Coefficients'][2]))
print("\n")
print("Compared with the standard room, a business room is associated with an RMB {:.2f} increase in the hotel room price.".format(df_coeff['Coefficients'][3]))
print("\n")
print("Compared with the standard room, a luxury room is associated with an RMB {:.2f} increase in the hotel room price.".format(df_coeff['Coefficients'][4]))

One unit of increase in average rating is associated with an RMB 165.03 increase in the hotel room price.


Compared with City 3, a hotel in City 1 is associated with an RMB 151.47 decrease in the hotel room price.


Compared with City 3, a hotel in City 2 is associated with an RMB 214.76 increase in the hotel room price.


Compared with the standard room, a business room is associated with an RMB 227.92 increase in the hotel room price.


Compared with the standard room, a luxury room is associated with an RMB 836.51 increase in the hotel room price.


More generally, one should be very careful when interpreting a linear regression model. As discussed above, a linear regression tries to estimate the conditional expectation of the outcome $Y$, given the feature vector $X$, i.e.,

$$f(X):=\mathbb E[Y|X],$$

assuming that $f(X)$ is linear in $X=(X_1,X_2,...,X_p)$. If the estimation result is

$$Y\approx \hat Y=\hat f(X)=\hat\beta_0+\sum_{j=1}^p\hat\beta_jX_{j}$$

Then, we have the following **direct** interpretations of the estimated parameters:

- $\hat\beta_0$: The estimated expected outcome when all the features are zero.
- $\hat\beta_j$: The change in the estimated expected outcome $\hat Y$ for a one unit change in feature $X_j$, holding all other features constant.
- **Association/correlation**: A one unit change in $X_j$ is associated with a $\hat\beta_j$ unit change in $Y$.

The above three interpretations are somehow "free-lunch" in the sense that you can interpret any linear regression model once it is built. However, we cannot make a stronger claim without further validation.


If we **fully** believe our **predictive** model (i.e., the model reasonably captures the ground-truth of how the data is generated), some sharper interpretations can be made:

- **Prediction**: Given a feature vector $(X_1,X_2,...,X_p)$, the predicted value of $Y$ is $\hat\beta_0+\sum_{j=1}^p\hat\beta_jX_j$.

Such a prediction interpretation can be made if the model has a strong predictive power, i.e., validated by a high out-of-sample $R^2$. Otherwise, it is not appropriate to predict the value of $Y$ with $\hat\beta_0+\sum_{j=1}^p\hat\beta_jX_j$.

Finally, if the variation in $X_j$ is **exogenous** (the exact definition of which will be introduced once we study Module 2, Causal Inference), we can even have a causal interpretation:

- **Causality**: A unit increase in $X_j$ causes a $\hat\beta_j$ unit increase in $Y$.

In most cases, such a strong causality interpretation is problematic, due to the notorious issue of endogeneity. So be very careful when you tend to make such a claim. We will discuss in detail when such a statement is possible and when it is not OK later in this course.

## 5.3. Correlation, Collinearity, and Identifiability

It may be possible that some features may have strong correlations. In this case, they will contain similar information in predicting the outcome $Y$. So it is advised to remove one of the two features that have a strong positive or negative correlation. 

Let us first formally define correlation. Consider a data set $\mathcal D=\{(X_i,Y_i):1\le i\le n,X_i\in\mathbb R^p,Y_i\in\mathbb R\}$. The correlation between feature $j$ and feature $l$ is defined by:

$$\rho_{jk}=\frac{\sum_{i=1}^n(X_{ij}-\bar X_j)(X_{ik}-\bar X_k)}{\sqrt{\sum_{i=1}^n(X_{ij}-\bar X_j)^2}\sqrt{\sum_{i=1}^n(X_{ik}-\bar X_k)^2}}\in[-1,1],$$
where $\bar X_j=\frac{1}{n}\sum_{i=1}^nX_{ij}$ is the average of feature $j$ and $\bar X_k=\frac{1}{n}\sum_{i=1}^nX_{ik}$ is the average of feature $l$. The bound $-1\le \rho_{jk}\le 1$ follows from the [Cauchy-Schwarz inequality](https://en.wikipedia.org/wiki/Cauchy%E2%80%93Schwarz_inequality). The interpretation of $\rho_{jk}$ can be summarized as follows:

- $\rho_{jk}>0$: Positive relationship between feature $j$ and feature $k$.
- $\rho_{jk}=1$: Perfect positive relationship between feature $j$ and feature $k$.
- $\rho_{jk}=0$: "No" relationship between feature $j$ and feature $k$.
- $\rho_{jk}<0$: Negative relationship between feature $j$ and feature $k$.
- $\rho_{jk}=-1$: Perfect negative relationship between feature $j$ and feature $k$.

Let's now compute the correlations between different features for the hotel room price prediction problem.

In [35]:
# Compute the correlations between each numeric variable pairs:
df.corr()

is_within_two_years  num_rating  rating_location  \
is_within_two_years             1.000000    0.011252         0.008735   
num_rating                      0.011252    1.000000         0.004143   
rating_location                 0.008735    0.004143         1.000000   
rating_service                 -0.008219   -0.014635        -0.014766   
rating_facility                 0.001825    0.004774        -0.011499   
POI_firm                        0.020523   -0.003188         0.013123   
POI_school                      0.023396   -0.025038         0.003861   
POI_travel                      0.017503   -0.002409         0.006659   
price                           0.135248   -0.012246        -0.017078   
a_rating                        0.001556   -0.001902         0.489350   
c1                              0.013248    0.005330        -0.001835   
c2                             -0.021637    0.003724        -0.002985   
rb                             -0.005482   -0.004200         0.010625   
rl                             -0.015474   -0.007291        -0.032144   
n_a_rating                     -0.001556    0.001902        -0.489350   
a_rating*rb                    -0.005600   -0.004015         0.016229   
a_rating*rl                    -0.015360   -0.007081        -0.026784   
log_price                       0.138756   -0.015913        -0.021210   

                     rating_service  rating_facility  POI_firm  POI_school  \
is_within_two_years       -0.008219         0.001825  0.020523    0.023396   
num_rating                -0.014635         0.004774 -0.003188   -0.025038   
rating_location           -0.014766        -0.011499  0.013123    0.003861   
rating_service             1.000000        -0.010247 -0.015714    0.020202   
rating_facility           -0.010247         1.000000 -0.004317    0.020519   
POI_firm                  -0.015714        -0.004317  1.000000    0.031440   
POI_school                 0.020202         0.020519  0.031440    1.000000   
POI_travel                 0.000015         0.000025 -0.017128   -0.002839   
price                      0.017524        -0.007247 -0.007481   -0.007216   
a_rating                   0.491515         0.704451 -0.004417    0.026859   
c1                        -0.010568         0.017584 -0.002134   -0.007090   
c2                         0.007063        -0.024783  0.011666   -0.007520   
rb                        -0.002005         0.012469 -0.015719    0.031700   
rl                         0.003863        -0.021638 -0.004139    0.034912   
n_a_rating                -0.491515        -0.704451  0.004417   -0.026859   
a_rating*rb                0.003624         0.020697 -0.015998    0.031956   
a_rating*rl                0.009261        -0.014346 -0.004164    0.035536   
log_price                  0.022562        -0.008501 -0.006520   -0.012549   

                     POI_travel     price  a_rating        c1        c2  \
is_within_two_years    0.017503  0.135248  0.001556  0.013248 -0.021637   
num_rating            -0.002409 -0.012246 -0.001902  0.005330  0.003724   
rating_location        0.006659 -0.017078  0.489350 -0.001835 -0.002985   
rating_service         0.000015  0.017524  0.491515 -0.010568  0.007063   
rating_facility        0.000025 -0.007247  0.704451  0.017584 -0.024783   
POI_firm              -0.017128 -0.007481 -0.004417 -0.002134  0.011666   
POI_school            -0.002839 -0.007216  0.026859 -0.007090 -0.007520   
POI_travel             1.000000  0.157771  0.003389  0.002125 -0.006992   
price                  0.157771  1.000000 -0.004938 -0.320469  0.378030   
a_rating               0.003389 -0.004938  1.000000  0.006303 -0.015662   
c1                     0.002125 -0.320469  0.006303  1.000000 -0.613129   
c2                    -0.006992  0.378030 -0.015662 -0.613129  1.000000   
rb                     0.018963 -0.283566  0.013272 -0.015550 -0.008373   
rl                    -0.003107  0.832957 -0.029759  0.004562  0.024966   
n_a_rating            -0.

As we can see, the average rating has a high positive correlation with the facility rating $\rho=0.704$. Let's see what will happen if we have both **rating_facility** and **a_rating** as features.

In [25]:
X2 = np.array(df[['a_rating','rating_facility','c1','c2','rb','rl']]).reshape([6125,6])

#Split the data into training and testing sets.

X2_train, X2_test, y2_train, y2_test = tr_te_split(X2, y, test_size=0.3)


# Fit a linear regression model with intercept (\beta_0)
reg2 = LinReg(fit_intercept=True).fit(X2_train, y2_train)

# Store the coefficients
df_coeff2 = pd.DataFrame({
    'Features': ['a_rating','rating_facility','c1','c2','rb','rl'], 
    'Coefficients': reg2.coef_
})
#The estimated coefficients
df_coeff2

Features  Coefficients
0         a_rating    169.446192
1  rating_facility      2.197955
2               c1   -146.238087
3               c2    210.612289
4               rb    226.836446
5               rl    836.509548

In [26]:
# The estimated coefficients without rating_facility
df_coeff

Features  Coefficients
0  a_rating    165.032013
1        c1   -151.465499
2        c2    214.756788
3        rb    227.923234
4        rl    836.511011

In [27]:
print("The out-of-sample R-squared of the linear regression model without rating_facility is:", reg1.score(X1_test,y1_test))
print("\n")
print("The out-of-sample R-squared of the linear regression model with rating_facility is:", reg2.score(X2_test,y2_test))

The out-of-sample R-squared of the linear regression model without rating_facility is: 0.8873617706600899


The out-of-sample R-squared of the linear regression model with rating_facility is: 0.8864265434182033


We remark that, due to the uncertainty in splitting the data into training and testing sets. The estimated coefficients and the $R^2$ values may be different for different trials. Based on intuition and on the high correlation between **rating_facility** and **a_rating**, we consider the model without rating_facility as a more suitable one. We do not have a strong reason to distinguish the rating on facility from other ratings. 

An extreme case of strongly correlated features is the issue of collinearity, in which one column of $X$ can be expressed as a linear combination of other columns. In this case, some feature(s) are redundant and should be directly dropped. Collinearity will result in the issue of non-identifiability, which refers to the situation where there exist multiple $\hat\beta$ with which the sum of squared errors is minimized.

Let's see an example of collinearity and non-identifiability where we add a new column **n_a_rating=-a_rating**.

In [18]:
# Create a new column with collinearity.
df['n_a_rating'] = - df['a_rating']

X3 = np.array(df[['a_rating','n_a_rating','c1','c2','rb','rl']]).reshape([6125,6])

#Split the data into training and testing sets.

X3_train, X3_test, y3_train, y3_test = tr_te_split(X3, y, test_size=0.3)

# Fit a linear regression model with intercept (\beta_0)
reg3 = LinReg(fit_intercept=True).fit(X3_train, y3_train)
# Store the coefficients
df_coeff3 = pd.DataFrame({
    'Features': ['a_rating','n_a_rating','c1','c2','rb','rl'], 
    'Coefficients': reg3.coef_
})
#The estimated coefficients
df_coeff3

Features  Coefficients
0    a_rating     83.565048
1  n_a_rating    -83.565048
2          c1   -151.486327
3          c2    210.134907
4          rb    229.520462
5          rl    836.845339

In the case where $p\approx n$ and $n$ is large, this is called the high-dimensional regime. In particular, if $p\ge n-1$, a linear regression model can perfectly fit the data. But such a model is generally poor-performing on unseen data because the model overfits data. If $p\ge n$, the linear regression is non-identificable in general.

## 5.4. Beyond Linearity

It is possible that the relationship between the outcome $Y$ and the features $X$ are inherently non-linear. In this case, it is recommended that you include some higher order terms such as $(X_j)^q$ for some $q\ge2$ or cross terms $X_jX_k$ as the features in linear regression. What to include may need some context and domain knowledge. In particular, if the value of one feature will affect the slope of another, we need the interaction term to be included in the model. For example, for different room types, the slope of average rating may be different. Therefore, we include the interaction terms **a_rating*rb** and **a_rating*rl**. The model is specified as:

$$\mbox{price}_i\approx \hat\beta_0+\hat\beta_1*\mbox{a_rating}_i+\hat\beta_2*\mbox{a_rating}_i*\mbox{rb}_i+\hat\beta_3*\mbox{a_rating}_i*\mbox{rl}_i+\hat\beta_4*\mbox{c1}_i+\hat\beta_5*\mbox{c2}_i+\hat\beta_5*\mbox{rb}_i+\hat\beta_5*\mbox{rl}_i$$

A key observation from the linear regression model with cross terms is that:

- If rb=rl=0 (the room type is standard), the slope of a_rating is $\hat\beta_1$.
- If rb=1 and rl=0 (the room type is business), the slope of a_rating is $\hat\beta_1+\hat\beta_2$.
- If rb=0 and rl=1 (the room type is luxury), the slope of a_rating is $\hat\beta_1+\hat\beta_3$.

Below we build the linear regression model with cross terms.

In [37]:
# Computing the interaction terms
df['a_rating*rb'] = df['a_rating']*df['rb']
df['a_rating*rl'] = df['a_rating']*df['rl']


X4 = np.array(df[['a_rating','a_rating*rb','a_rating*rl','c1','c2','rb','rl']]).reshape([6125,7])

#Split the data into training and testing sets.

X4_train, X4_test, y4_train, y4_test = tr_te_split(X4, y, test_size=0.3)

# Fit a linear regression model with intercept (\beta_0)
reg4 = LinReg(fit_intercept=True).fit(X4_train, y4_train)
# Store the coefficients
df_coeff4 = pd.DataFrame({
    'Features': ['a_rating','a_rating*rb','a_rating*rl','c1','c2','rb','rl'], 
    'Coefficients': reg4.coef_
})
#The estimated coefficients
df_coeff4

Features  Coefficients
0     a_rating     72.316366
1  a_rating*rb     90.093340
2  a_rating*rl    214.538302
3           c1   -152.805141
4           c2    207.781132
5           rb   -170.881016
6           rl   -122.532311

In [29]:
# Out-of-sample R-squared for the model with cross terms
print("The out-of-sample R-squared of the linear regression model with cross-terms is:", reg4.score(X4_test,y4_test))

The out-of-sample R-squared of the linear regression model with cross-terms is: 0.8900429003155855


Based on the estimation results of the new model with feature interactions, we have the following interpretations:

In [30]:
print("If the room type is standard (rb=rl=0), one unit of increase in average rating is associated with an RMB {:.2f} increase in room price.".format(df_coeff4['Coefficients'][0])) 
print("\n")
print("If the room type is business (rb=1 and rl=0), one unit of increase in average rating is associated with an RMB {:.2f} increase in room price.".format(df_coeff4['Coefficients'][0]+df_coeff4['Coefficients'][1])) 
print("\n")
print("If the room type is luxury (rb=0 and rl=1), one unit of increase in average rating is associated with an RMB {:.2f} increase in room price.".format(df_coeff4['Coefficients'][0]+df_coeff4['Coefficients'][2])) 

If the room type is standard (rb=rl=0), one unit of increase in average rating is associated with an RMB 142.84 increase in room price.


If the room type is business (rb=1 and rl=0), one unit of increase in average rating is associated with an RMB 121.38 increase in room price.


If the room type is luxury (rb=0 and rl=1), one unit of increase in average rating is associated with an RMB 261.31 increase in room price.


Hence, for different room types, the association between room price and average rating is different. To decide whether interaction terms should be included, we have the following rule-of-thumbs:

- Include interactions with a feature with large coefficient in a standard linear regression without interactions.
- Include interactions with a feature describing groups of data (e.g., hotel room type).

## 5.5. Logarithmic Transformations

In a lot of contexts, the outcome is positive (e.g., height, counts, prices, revenues, etc.), but the linear regression model may predict that the value of $Y$ is negative for some feature vector $X$. One approach to address this issue is to take the logarithmic transformation of the data before applying the linear regression:

$$\mbox{log}(Y_i)\approx \hat\beta_0+\sum_{j=1}^p\hat\beta_jX_{ij}$$

Taking exponential, we have

$$Y_i\approx \mbox{exp}(\hat\beta_0)\mbox{exp}(\hat\beta_1X_{i1})...\mbox{exp}(\hat\beta_pX_{ip})$$

By the Taylor expansion, we have $\mbox{exp}(\hat\beta_j)\approx 1+\hat\beta_j$, so

$$\exp(\hat\beta_j X_{ij})=(\exp(\hat\beta_j))^{X_{ij}}\approx (1+\hat\beta_j)^{X_{ij}}$$

and

$$Y_i\approx\mbox{exp}(\hat\beta_0)(1+\hat\beta_1)^{X_{i1}}(1+\hat\beta_2)^{X_{i2}}...(1+\hat\beta_p)^{X_{ip}}$$

Hence, we can interpret the model as

- One unit of change in $X_j$ is suggesting a **proportional** change of $\hat\beta_j$ in the fitted value $\hat Y$.
- If both the outcome and the feature are logged, the linear regression coefficient gives proportional changes in the fitted outcome value associated with the proportional changes in the feature.

We now give a demonstration using the hotel room price case by taking the log transformation of room price.

In [31]:
# Computing the log of price
df['log_price'] = np.log(df['price'])


X5 = np.array(df[['a_rating','a_rating*rb','a_rating*rl','c1','c2','rb','rl']]).reshape([6125,7])
y5 = np.array(df['log_price'])
#Split the data into training and testing sets.

X5_train, X5_test, y5_train, y5_test = tr_te_split(X5, y5, test_size=0.3)

# Fit a linear regression model with intercept (\beta_0)
reg5 = LinReg(fit_intercept=True).fit(X5_train, y5_train)
# Store the coefficients
df_coeff5 = pd.DataFrame({
    'Features': ['a_rating','a_rating*rb','a_rating*rl','c1','c2','rb','rl'], 
    'Coefficients': reg5.coef_
})
#The estimated coefficients
df_coeff5

Features  Coefficients
0     a_rating      0.151051
1  a_rating*rb     -0.003861
2  a_rating*rl      0.030226
3           c1     -0.191765
4           c2      0.214731
5           rb      0.335616
6           rl      0.712395

In [32]:
# Out-of-sample R-squared for the model with log_price as the outcome variable
print("The out-of-sample R-squared of the linear regression model with log_price is:", reg5.score(X5_test,y5_test))

The out-of-sample R-squared of the linear regression model with log_price is: 0.8145685719047557


Thus, the model with the logged price being the label does not perform as well as the model with price itself being the outcome variable.

## 5.6. Concluding Remarks

Linear regression is arguably the **simplest** supervised learning model for regression (predicting a continuous label). It assumes a linear relationship between the label and the features and the estimation seeks to minimize the MSE (mean-squared error) of all training observations. We use the in-sample $R^2$ (which is lower-bounded by 0 and upper-bounded by 1) to evaluate the goodness-of-fit of the linear regression model. As for the prediction accuracy, the basic metric is the out-of-sample $R^2$, which compares the squared error of the linear regression model and that of the naive prediction without using any features. Linear regression has great interpretability, but you should be cautious not to confuse correlation with causation. 